src/Storages/VirtualColumnUtils.cpp

```diff
@@ -214,26 +224,46 @@ static void addPathAndFileToVirtualColumns(Block & block, const String & path, s
         block.getByName("_file").column->assumeMutableRef().insert(file);
     }
 
+    std::unordered_map<std::string, std::string> keys;
+    if (context->getSettingsRef().use_hive_partitioning)
+        keys = parseHivePartitioningKeysAndValues(path);
+
+    for (const auto & virt_column : virtual_columns)
+    {
+        auto it = keys.find(virt_column.name);
+        if (it != keys.end())
+        {
+            if (!block.has(virt_column.name))
+                block.insert({virt_column.type->createColumn(), virt_column.type, virt_column.name});
```

Comment:
This looks unnecessary. We should've insert all required virtual columns in block in `createPathAndFileFilterDAG`

Reply:
`createPathAndFileFilterDAG()` is created in the `FileIterator` before the paths are parsed so it now fails because required columns are absent

Reply:
Ah, then we should inesrt these virtual columns in `getFilterByPathAndFileIndexes` together with `_path` and `_file`. It's not good to add it here in `addFilterDataToVirtualColumns`

Reply:
done